python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -987,6 +988,20 @@ def automation_conditions_by_key(self) -> Mapping[AssetKey, AutomationCondition]
             if spec.automation_condition
         }
 
+    @functools.lru_cache
+    def get_upstream_input_keys(self, keys: frozenset[AssetKey]) -> AbstractSet[AssetKey]:
+        """Returns keys that are directly upstream of the provided keys and are inputs of this asset."""
+        direct_upstreams = {dep.asset_key for key in keys for dep in self._specs_by_key[key].deps}
+        return direct_upstreams & set(self.keys_by_input_name.values())
+
+    @functools.lru_cache
+    def get_checks(self, keys: frozenset[AssetKey]) -> AbstractSet[AssetCheckKey]:
+        """Returns checks defined on this AssetsDefinition for the provided keys."""
+        check_keys = {
+            check_spec.key for check_spec in self.node_check_specs_by_output_name.values()
+        }
+        return {key for key in check_keys if key.asset_key in keys}
+
```

Comment:
@dpeng817 I added these as more "general purpose" utility functions that have the benefit of being cached for perf